scratch_space/lib/src/scratch_space.dart

```diff
@@ -80,30 +83,31 @@ class ScratchSpace {
       throw new StateError('Tried to use a deleted ScratchSpace!');
     }
 
-    var futures = <Future>[];
-    for (var id in assetIds) {
-      // Always track `id` as a dependency, we do this by just calling
-      // `canRead` for now.
-      //
-      // ignore: unawaited_futures
-      reader.canRead(id);
-      var file = fileFor(id);
-      if (file.existsSync()) {
-        var pending = _pendingWrites[id];
-        if (pending != null) futures.add(pending);
-      } else {
-        file.createSync(recursive: true);
-        var done = reader
-            .readAsBytes(id)
-            .then((bytes) =>
-                _descriptorPool.withResource(() => file.writeAsBytes(bytes)))
-            .whenComplete(() {
-          _pendingWrites.remove(id);
-        });
-        _pendingWrites[id] = done;
-        futures.add(done);
+    var futures = assetIds.map((id) async {
+      var digest = await reader.digest(id);
+      var existing = _digests[id];
+      if (digest == existing) {
+        await _pendingWrites[id];
+        return;
       }
-    }
+      _digests[id] = digest;
+
+      try {
+        await _pendingWrites.putIfAbsent(
+            id,
+            () => _descriptorPool.withResource(() async {
+                  var file = fileFor(id);
+                  if (await file.exists()) {
+                    await file.delete();
+                  }
+                  await file.create(recursive: true);
+                  await file.writeAsBytes(await reader.readAsBytes(id));
+                }));
+      } finally {
+        _pendingWrites..remove(id);
```

Comment:
is the `..` necessary here?

Reply:
It just gets rid of the unawaited futures hint